core/org.wso2.carbon.core/src/main/java/org/wso2/carbon/core/clustering/hazelcast/HazelcastClusterMessageAsyncListener.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.core.clustering.hazelcast;
+
+import com.hazelcast.core.Message;
+import com.hazelcast.core.MessageListener;
+import org.apache.axis2.clustering.ClusteringFault;
+import org.apache.axis2.clustering.ClusteringMessage;
+import org.apache.axis2.context.ConfigurationContext;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+/**
+ * Listener which should be used for non-guaranteed delivery mode.
+ * e.g. Cache Invalidation Messages.
+ */
+public class HazelcastClusterMessageAsyncListener implements MessageListener<AsyncClusteringMessage> {
```

Comment:
@ruwanta Can you let me know why this is named as an async listener? How is this different from other message listeners?

Reply:
Will be renamed to Idempotent MessageListener , as there is no harm replaying cache invalidation message.

Reply:
We decided to have only one message listener and remove the replaying part right? So the only big change would be to remove the sent and received message buffers from the cluster messaging code we have right now, and we don't need to introduce new listeners etc.

Reply:
+1

Reply:
@afkham on the second thought, I believe we need to keep current behavior of not sending the same message or nor processing the same message as it is. (Which means retaining the send and received buffers for all other non-cache related flows). Still some products depend on things like Dep-Sync.